pkg/query/union_test.go

```diff
@@ -461,3 +458,179 @@ func TestUnionDeduplicationBugFix(t *testing.T) {
 		require.Equal("alice", paths[0].Subject.ObjectID)
 	})
 }
+
+func TestUnionIteratorCaveatCombination(t *testing.T) {
+	t.Parallel()
+
+	require := require.New(t)
+
+	// Create test context
+	ctx := &Context{
+		Context:  t.Context(),
+		Executor: LocalExecutor{},
+	}
+
+	t.Run("CombineTwoCaveats_OR_Logic", func(t *testing.T) {
+		t.Parallel()
+
+		// Both paths have different caveats - should combine with OR logic
+		pathWithCaveat1 := MustPathFromString("document:doc1#viewer@user:alice")
+		pathWithCaveat1.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "caveat1",
+				},
+			},
+		}
+
+		pathWithCaveat2 := MustPathFromString("document:doc1#viewer@user:alice")
+		pathWithCaveat2.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "caveat2",
+				},
+			},
+		}
+
+		iter1 := NewFixedIterator(pathWithCaveat1)
+		iter2 := NewFixedIterator(pathWithCaveat2)
+
+		union := NewUnion()
+		union.addSubIterator(iter1)
+		union.addSubIterator(iter2)
+
+		relSeq, err := ctx.Check(union, NewObjects("document", "doc1"), NewObject("user", "alice").WithEllipses())
+		require.NoError(err)
+
+		rels, err := CollectAll(relSeq)
+		require.NoError(err)
+
+		require.Len(rels, 1, "Union should deduplicate to one relation")
+		require.NotNil(rels[0].Caveat, "Result should have combined caveat")
+		// The combination logic should preserve one of the caveats (implementation detail)
```

Comment:
Because it's evaluating one of the caveats in the process? Otherwise wouldn't you need to preserve both caveats, because one (or more) could evaluate to false and the overall expression could still be true?

Reply:
Extending this test (and others) to exemplify the combine process